judoscale/celery/collector.py

```diff
@@ -0,0 +1,88 @@
+import json
+from threading import Thread
+from typing import List, Optional, Set
+
+from celery import Celery
+from kombu import Connection
+from redis import Redis
+
+from judoscale.core.config import Config
+from judoscale.core.logger import logger
+from judoscale.core.metric import Metric
+from judoscale.core.metrics_collectors import JobMetricsCollector
+
+
+class TaskSentHandler(Thread):
+    def __init__(
+        self,
+        collector: "CeleryMetricsCollector",
+        connection: Connection,
+        *args,
+        **kwargs,
+    ):
+        self.collector = collector
+        self.connection = connection
+        super().__init__(*args, daemon=True, **kwargs)
+
+    def task_sent(self, event):
+        self.collector.queues.add(event["queue"])
+
+    def run(self):
+        logger.debug("Starting TaskSentHandler")
+        recv = self.collector.broker.events.Receiver(
+            self.connection,
+            handlers={"task-sent": self.task_sent},
+        )
+        recv.capture(wakeup=False)
+
+
+class CeleryMetricsCollector(JobMetricsCollector):
+    def __init__(self, config: Config, broker: Celery):
+        super().__init__(config=config)
+
+        self.broker = broker
+        connection = self.broker.connection_for_read()
+        if connection.transport.driver_name != "redis":
+            raise NotImplementedError(
+                f"Unsupported broker: {connection.transport.driver_name}"
+            )
+
+        self.redis: Redis = connection.channel().client
+        self.queues: Set[str] = set()
+        self.task_sent_handler = TaskSentHandler(self, connection)
+        logger.debug(f"Redis is at {self.redis.connection_pool}")
+
+        system_queues = {"unacked", "unacked_index"}
+        user_queues = {
+            q.decode("utf-8") if type(q) is bytes else q
+            for q in self.redis.scan_iter(match="[^_]*", _type="list")
+        }
+        logger.debug(f"Found initial queues: {list(user_queues)}")
+        self.queues = user_queues - system_queues
+        self.task_sent_handler.start()
+
+    def oldest_task(self, queue: str) -> Optional[dict]:
+        """
+        Get the oldest task from the queue.
+        """
+        try:
+            if payload := self.redis.lindex(queue, -1):
+                return json.loads(payload)
+        except Exception as e:
+            logger.warning(f"Unable to get a task from queue: {queue}", exc_info=e)
+        return None
+
+    def collect(self) -> List[Metric]:
+        metrics = []
+        if not self.should_collect:
+            return metrics
+
+        logger.debug(f"Collecting metrics for queues {list(self.queues)}")
```

Comment:
Found an issue here that I don't think I was clear about before. Even when nothing is pending in a given queue, we still want to report queue time for all known queues. In other words, reporting a queue time of 0 is better than not reporting the queue at all. This way Judoscale knows that the queue has been cleared, and it's not just waiting for the next metric.

In my testing, I've enqueued and processed jobs on the "low" and "high" queues. The queues are now empty. Here's what I see in the logs:

```
8:10:12 AM worker.1      |  DEBUG - [judoscale] Collecting metrics for queues ['low', 'high']
8:10:12 AM web.1         |  DEBUG - [judoscale] Collecting metrics for queues ['low', 'high']
8:10:12 AM worker_high.1 |  DEBUG - [judoscale] Collecting metrics for queues ['high', 'low']
8:10:12 AM worker.1      |  DEBUG - [judoscale] Posting 0 metrics from 6036 to Judoscale adapter API https://requestinspector.com/inspect/judoscale-celery-adamlogic/v3/reports
8:10:12 AM worker_high.1 |  DEBUG - [judoscale] Posting 0 metrics from 6037 to Judoscale adapter API https://requestinspector.com/inspect/judoscale-celery-adamlogic/v3/reports
8:10:12 AM web.1         |  DEBUG - [judoscale] Posting 0 metrics from 6035 to Judoscale adapter API https://requestinspector.com/inspect/judoscale-celery-adamlogic/v3/reports
```

Ideally, I still want each of these processes "posting 2 metrics" instead of "posting 0 metrics". They should be posting a 0 queue time for each of the known queues.

Does this make sense?

Reply:
Yep, that makes sense. I've made the change to report latency on empty queues as well.